c-deps/libroach/stack_trace.cc

```diff
@@ -0,0 +1,271 @@
+  if (stack == nullptr) {
+    return;
+  }
+  stack->depth = backtrace(stack->addr, kMaxDepth);
+  stack->Ack();
+}
+
+int SignalThread(pid_t pid, pid_t tid, uid_t uid, int signum, sigval payload) {
+  // Similar to pthread_sigqueue(), but usable with a tid since we
+  // don't have a pthread_t.
+  siginfo_t info;
+  memset(&info, 0, sizeof(info));
+  info.si_signo = signum;
+  info.si_code = SI_QUEUE;
+  info.si_pid = pid;
+  info.si_uid = uid;
+  info.si_value = payload;
+  return syscall(SYS_rt_tgsigqueueinfo, pid, tid, signum, &info);
+}
+
+bool InitThreadDumper() {
+  struct sigaction action;
+  memset(&action, 0, sizeof(action));
+  action.sa_sigaction = InternalHandler;
+  // Set SA_RESTART so that supported syscalls are automatically restarted if
+  // interrupted by the stacktrace collection signal.
+  action.sa_flags = SA_ONSTACK | SA_RESTART | SA_SIGINFO;
+  return sigaction(kStackTraceSignal, &action, nullptr) == 0;
+}
+
+int64_t NowMillis() {
+  timeval tv;
+  gettimeofday(&tv, NULL);
+  return (tv.tv_sec * 1000) + (tv.tv_usec / 1000);
+}
+
+}  // namespace
+
+std::string DumpThreadStacks() {
+  auto tids = ListThreads();
+  InitThreadDumper();
+
+  // Create a pipe on which threads can send acks after they finish
+  // writing their stacktrace.
+  int pipe_fd[2];
+  if (pipe(pipe_fd) == -1) {
+    return "unable to create pipe";
+  }
+
+  // Signal all threads to write their stack trace in a pre-allocated
+  // area. Note that some threads might have died by now, so
+  // signalling them will fail.
+  std::vector<std::unique_ptr<ThreadStack>> stacks;
+  const auto pid = getpid();
+  const auto uid = getuid();
+  for (auto tid : tids) {
+    const uint64_t blocked = BlockedSignals(tid);
+    if ((blocked & (1ULL << kStackTraceSignal)) != 0) {
+      // The thread is blocking receipt of our signal, so don't bother
+      // sending it.
+      continue;
+    }
+
+    std::unique_ptr<ThreadStack> stack(new ThreadStack(tid, pipe_fd[1]));
+    union sigval payload;
+    payload.sival_ptr = stack.get();
+    if (SignalThread(pid, tid, uid, kStackTraceSignal, payload) == 0) {
+      stacks.push_back(std::move(stack));
+    }
+  }
+
+  // Set operations on pipe_fd[0] to be non-blocking. This is
+  // important if the poll() on this fd returns, but the subsequent
+  // read block.
+  int flags = fcntl(pipe_fd[0], F_GETFL, 0);
+  fcntl(pipe_fd[0], F_SETFL, flags | O_NONBLOCK);
+
+  // Wait for all the acks, timing out after 5 seconds.
+  auto end = NowMillis() + 5000;
+  for (int acks = 0; acks < stacks.size(); ) {
+    pollfd pollfds[1];
+    pollfds[0].fd = pipe_fd[0];
+    pollfds[0].events = POLLIN;
+    pollfds[0].revents = 0;
+    auto timeout = end - NowMillis();
+    if (timeout <= 0) {
+      break;
+    }
+    auto ret = poll(pollfds, 1, int(timeout));
+    if (ret == -1) {
+      continue;
+    }
+    if (ret == 0) {
+      // We timed out before reading all of the stacks.
```

Comment:
Are we returning partial info in this case? Are we marking it as partial?

Reply:
Done. See the code below where we indicate `(no response)` for a thread which didn't return a stack.